services/venice-controller/src/main/java/com/linkedin/venice/controller/StoreBackupVersionCleanupService.java

```diff
@@ -249,8 +260,19 @@ protected boolean cleanupBackupVersion(Store store, String clusterName) {
       stats.recordBackupVersionMismatch();
       return false;
     }
-    List<Version> readyToBeRemovedVersions =
-        versions.stream().filter(v -> v.getNumber() < currentVersion).collect(Collectors.toList());
+    int repushSourceVersion = NON_EXISTING_VERSION;
+    for (Version version: versions) {
+      if (version.getRepushSourceVersion() > NON_EXISTING_VERSION) {
+        repushSourceVersion = version.getRepushSourceVersion();
+      }
+    }
+    int finalRepushSourceVersion = repushSourceVersion;
+    List<Version> readyToBeRemovedVersions = versions.stream()
+        .filter(
+            v -> finalRepushSourceVersion > NON_EXISTING_VERSION
+                ? v.getNumber() == finalRepushSourceVersion
```

Comment:
So here, we either cleanup the source version or true backup version, which passes the delay, not both, right?
It is fine to have such logic, and please add some javadoc to describe such desired behavior, otherwise, it is a little confusing.

Reply:
Sure will add a javadoc, its basically first deletes the repush source version, then in later iteration of the loop after threshold wait time (eg 7 days) it will delete the true backup version.